### ISSUE TITLE
Booting Linux

### ISSUE THREAD

**Issue by natinusala** [DavidBuchanan314/NXLoader#9]
This is a discussion about what's needed to boot Linux from NXLoader, taking [TegraRcmSmash](https://github.com/rajkosto/TegraRcmSmash) as an example.

- Be able to change or remove the relocator payload (no relocator is needed to boot Linux) ([line 761](https://github.com/rajkosto/TegraRcmSmash/blob/master/Smasher.cpp))
- Be able to send "sections" (`coreboot.rom` file) ([lines 896 to 1042](https://github.com/rajkosto/TegraRcmSmash/blob/master/Smasher.cpp))

There is no need to use imx_usb with [my method of booting Linux](https://github.com/natinusala/switch-linux), so I suggest we use this one and not bother with imx.

The only files needed will be `cbfs.bin` and `coreboot.rom`. As they will probably not be updated (unless we make a change in u-boot), I suggest we embed them in the app to have a one-click Linux boot.

**Comment by DavidBuchanan314** [comment]
Worth looking at this: https://github.com/MCMrARM/switch_linux_launcher

**Comment by DavidBuchanan314** [comment]
I think it should be possible to write a fusee payload that loads coreboot.rom from SD

**Comment by natinusala** [comment]
It should, however I'm afraid I lack the knowledge to do so. fusée gelée doesn't have a sdmmc driver yet, so it's not on our side (yet).

**Comment by natinusala** [comment]
I was able to port some of TegraRcmSmash, Shofel2 and switch_linux_launcher code into NXLoader. Here is what I come up with (no commit since it will break my PR) : https://pastebin.com/V7rZG0MA

Don't worry about the `BOOTING_LINUX` constant, it will be dropped in favor to user settings.

The current code doesn't work as I never recieve the `CBFS` request from coreboot (I don't receive anything actually).

This can mean two things : 
1. The method to read the request is wrong, or
2. The payload is not correctly executed

My guess is more on the second one. Even though the payload is of the same length as the one that the original shofel2 sends, I think that it may not be correctly sent and / or executed.

I think this has to be with the DMA buffers and the  `0x7000` constant. switch_linux_launcher uses similar code to NXLoader but with `0x6C68` instead of `0x7000`.

Something I also noticed is that all other implementations send the header first (first 680 bytes), and _then_ the payload. NXLoader sends everything at once. Maybe this messes the buffer alignment up ? Other implementations have a "ensure that we are on higher buffer" (fusée gelée, TegraRcmSmash) or "sanity check / throwing more data" (shofel2, switch_linux_launcher) that we don't have here.

**Comment by DavidBuchanan314** [comment]
In my implementation, the `low_buffer` keeps track of the current DMA buffer. fusee-launcher sends the whole payload at once.

I also had a go a while back, and ran into the same issue with never recieving the `CBFS` string. Have you tried using `switch_linux_launcher` on it's own? (I haven't, no time). I'm wondering whether maybe it just won't work on some devices.

rajkosto is working on getting fusee to bootstrap linux from the SD, so tbh I'd rather just wait for that, no need to implement a feature in the app which will soon become obsolete. He reckons he might get it done in the next couple of days. (fusee already has sdmmc support btw)

**Comment by natinusala** [comment]
If rajkosto is working on it then I agree that it's better to keep this app as minimal as possible.

I've tried `switch_linux_launcher`, CMake was not happy with `imx_usb_loader` so I just removed it altogether

And it works, so we must be doing something wrong here.

**Comment by DavidBuchanan314** [comment]
If I had to guess where the bug is, it's in the REAPURB part, because TBH I don't understand what it does very well.

`switch_linux_launcher` does things subtly differently (and probably correctly).

**Comment by natinusala** [comment]
Would it be that NXLoader breaks the USB connection by wrongly reaping the URB and thus it can never read the CBFS signal ?

**Comment by DavidBuchanan314** [comment]
Yeah, that sounds plausible.

**Comment by DavidBuchanan314** [comment]
Actually, looking a bit more closely, my usage of REAPURB is completely broken

**Comment by natinusala** [comment]
Should I bother fixing it to see if it allows us to boot Linux using CBFS ?

**Comment by gotwig** [comment]
yes pls :) any progress on this? and cant you run a console on android? you could port the same toolchain over to android, no?

**Comment by natinusala** [comment]
I suggest you try memloader, a fusee-launcher payload which allows to boot into Linux without sideloaded anything. If it works, this issue becomes irrelevant.
